### ISSUE TITLE
RLgraph: Robust incrementally testable reinforcement learning

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3094]
RLgraph: Robust, incrementally testable reinforcement learning<br>
<br>
https://ift.tt/2SJuYM6<br>
<br>
<br>
<br>
<div><section><h1>RLgraph: Robust, incrementally testable reinforcement learning</h1> <p>We introduce RLgraph (<a href="https://github.com/rlgraph/rlgraph">GitHub repo</a>), a RL framework decoupling logical component composition from deep learning backend and distributed execution. RLgraph brings rigorous management of internal and external state, inputs, devices, and dataflow to reinforcement learning.</p> <p><img alt="APEX learning on multiple GPUs" src="https://rlgraph.github.io/images/learning_combined.png"><span><strong>Left:</strong> Soft Actor Critic on Pendulum-v0 (10 seeds). <strong>Right:</strong> Multi-GPU Ape-X on Pong-v0 (10 seeds).</span></p> <p>Using RLgraph, developers combine high level components in a space-independent manner and define input spaces. RLgraph then builds a backend-independent component graph which can be translated to a TensorFlow computation graph, or executed in define by run mode via PyTorch (1.0+).</p> <p>The resulting agents can be trained locally embedded in an application like any other library (e.g. numpy), or using backends such as distributed TensorFlow, Ray, and Horovod. RLgraph currently implements variants of contemporary methods such as SAC (thanks to a contributor), IMPALA, APE-X, PPO.</p> <p><img alt="RLgraph stack" src="https://rlgraph.github.io/images/rlgraph_stack.png"></p> <p>This design has a number of advantages compared to many existing implementations which struggle to resolve the tension between prototyping, reusability across backends, and scalable execution:</p> <ul>
<li>Each component can be built and tested as its own graph - being able to systematically test arbitrary sub-graphs of complex algorithms drastically accelerates debugging and prototyping. Think of components as Sonnet-style objects but including API and device management, build system, separation of variable creation and computation logic.</li> <li>Backend and space-independent high-level logic enables faster exploration of new designs.</li> <li>Scalability: RLgraph agents can be arbitrarily executed e.g. on Ray using our Ray execution package, using distributed TensorFlow to explore end-to-end graphs, or using any distribution mechanism of choice.</li> <li>Maintenance: Extending existing frameworks often means either duplicating large amounts of code or intransparently piling on new learning heuristics into a single implementation. Neither is desirable. In RLgraph, heuristics are first-class citizens which are separately built and tested.</li> </ul>
<p>Overall, RLgraph targets an opinionated trade-off between enforcing strict interfaces and a rigorous build system, and exploration of new designs. Specifically, RLgraph is a framework meant for users that need to execute their research in practice in all kinds of deployment contexts. There is also a simple plug and play high level API (see below).</p> <h2>Architecture</h2> <p>Users interact with RLgraph agents through a front-end agent API very similar to TensorForce (which some of us also created/worked on). Internally, agents rely on a graph executor which serves requests against the component graph and manages backend-specific execution semantics. For example,the TensorFlowExecutor handles sessions, devices, summaries, placeholders, profiling, distributed TF servers, and timelines. The PyTorch executor in turn manages default tensor types, devices, or external plugins such as Horovod.</p> <p><img alt="RLgraph execution" src="https://rlgraph.github.io/images/rlgraph_graph_execution.png"></p> <p>This agent can be executed on external engines (e.g. using RLgraph’s <code>RayExecutors</code>) or be used like any other object in an application.</p> <h2>API examples</h2> <p>Most applied users will be able to rely on a comprehensive high-level API. We show how agents can be configured and executed seamlessly in local and distributed contexts:</p> <div><div><pre><code><span>from</span> <span>rlgraph.agents</span> <span>import</span> <span>agents</span> <span>from</span> <span>rlgraph.spaces</span> <span>import</span> <span>*</span> <span># Describe inputs via powerful space objects which manage batch and time ranks.</span> <span>states</span> <span>=</span> <span>IntBox</span><span>(</span> <span>low</span><span>=</span><span>0</span><span>,</span> <span>high</span><span>=</span><span>1024</span><span>,</span> <span>add_time_rank</span><span>=</span><span>True</span> <span>)</span> <span># Nested container spaces are used throughout RLgraph</span> <span># to describe state and perform shape inference.</span> <span>actions</span> <span>=</span> <span>Dict</span><span>(</span> <span>int_action</span><span>=</span><span>IntBox</span><span>(</span><span>low</span><span>=</span><span>0</span><span>,</span> <span>high</span><span>=</span><span>5</span><span>),</span> <span>bool_action</span><span>=</span><span>BoolBox</span><span>(),</span> <span>float_action</span><span>=</span><span>FloatBox</span><span>(</span><span>shape</span><span>=</span><span>(</span><span>3</span><span>,)</span> <span>)</span> <span># Create PPO agent</span> <span>agent</span> <span>=</span> <span>Agent</span><span>.</span><span>from_spec</span><span>(</span> <span>agent_config</span><span>,</span> <span>state_space</span><span>=</span><span>env</span><span>.</span><span>state_space</span><span>,</span> <span>action_space</span><span>=</span><span>env</span><span>.</span><span>action_space</span> <span>)</span> <span># Use agent locally, control behaviour with flags.</span> <span>actions</span> <span>=</span> <span>agent</span><span>.</span><span>get_actions</span><span>(</span><span>states</span><span>,</span> <span>use_exploration</span><span>=</span><span>use_exploration</span><span>,</span> <span>apply_preprocessing</span><span>=</span><span>True</span><span>)</span> <span># Batch observe multi-environment settings.</span> <span>agent</span><span>.</span><span>observe</span><span>(</span><span>states</span><span>,</span> <span>actions</span><span>,</span> <span>rewards</span><span>,</span> <span>terminals</span><span>,</span> <span>batched</span><span>=</span><span>True</span> <span>,</span> <span>env_id</span><span>=</span><span>"env_3"</span><span>)</span> <span># Updates by sampling from buffer after observing.</span> <span>loss</span> <span>=</span> <span>agent</span><span>.</span><span>update</span><span>()</span> <span># Update from an external batch which may contain </span> <span># arbitrary (non-terminal) sub-episode fragments from multiple environments,</span> <span># identified via sequence indices:</span> <span>agent</span><span>.</span><span>update</span><span>(</span><span>batch</span><span>=</span><span>dict</span><span>(</span><span>states</span><span>=</span><span>states</span><span>,</span> <span>actions</span><span>=</span><span>actions</span><span>,</span> <span>rewards</span><span>=</span><span>rewards</span><span>,</span> <span>terminals</span><span>=</span><span>terminals</span><span>,</span> <span>sequence_indices</span><span>=</span><span>sequence_indices</span><span>)</span> <span># Single-threaded, vectorized execution. </span> <span>env_spec</span> <span>=</span> <span>{</span><span>"type"</span><span>:</span> <span>"openai"</span><span>,</span> <span>"env_id"</span><span>:</span> <span>"CartPole-v0"</span><span>}</span> <span>worker</span> <span>=</span> <span>SingleThreadedWorker</span><span>(</span><span>env_spec</span><span>=</span><span>env_spec</span><span>,</span> <span>agent</span><span>=</span><span>agent</span><span>,</span> <span>num_envs</span><span>=</span><span>8</span><span>)</span> <span>stats</span> <span>=</span> <span>worker</span><span>.</span><span>execute_timesteps</span><span>(</span><span>100000</span><span>)</span> <span># Go large scale with Ray in 2 lines.</span> <span>ray_executor</span> <span>=</span> <span>SyncBatchExecutor</span><span>(</span><span>agent_config</span><span>,</span> <span>env_spec</span><span>)</span> <span>ray_stats</span> <span>=</span> <span>ray_executor</span><span>.</span><span>execute</span><span>(</span><span>steps</span><span>=</span><span>100000</span><span>)</span> <span># Use agent trained on Ray just as before.</span> <span>agent</span> <span>=</span> <span>ray_executor</span><span>.</span><span>local_agent</span> </code></pre></div></div> <p>Full <a href="https://github.com/rlgraph/rlgraph/tree/master/examples">example scripts</a> and configuration files can be found in the repository.</p> <h2>Components</h2> <p>Everything in RLgraph is a <code>Component</code>. An agent implements an algorithm via a root component containing sub-components like memories, neural networks, loss functions, or optimizers. Components interact with each other through API functions which impliitly create data-flow between components. That is, using the TensorFlow backend, RLgraph creates end-to-end static graphs by stitching together API calls and tracing them during the build procedure. For example, below is a simple API method to update the policy network:</p> <p><img alt="RLgraph API example" src="https://rlgraph.github.io/images/rlgraph_api_example.png"></p> <p>API method decorators wrap API functions to create end-to-end dataflow. RLgraph manages session, variable/internal states, devices, scopes, placeholders, nesting, time and batchranks for each component and its incoming and outgoing dataflow.</p> <p>We can build a component from a space and interact with it without needing to manually create tensors, input placeholders etc:</p> <div><div><pre><code><span>record_space</span> <span>=</span> <span>Dict</span><span>(</span> <span>states</span><span>=</span><span>dict</span><span>(</span><span>state1</span><span>=</span><span>float</span><span>,</span> <span>state2</span><span>=</span><span>float</span><span>),</span> <span>actions</span><span>=</span><span>dict</span><span>(</span><span>action1</span><span>=</span><span>float</span><span>,</span> <span>action2</span><span>=</span><span>IntBox</span><span>(</span><span>10</span><span>)),</span> <span># For scalar spaces, use Python primitives</span> <span>reward</span><span>=</span><span>float</span><span>,</span> <span>terminals</span><span>=</span><span>BoolBox</span><span>(),</span> <span>next_states</span><span>=</span><span>dict</span><span>(</span><span>state1</span><span>=</span><span>float</span><span>,</span> <span>state2</span><span>=</span><span>float</span><span>),</span> <span>add_batch_rank</span><span>=</span><span>True</span> <span>)</span> <span># Memory exposes insert, sample..methods</span> <span>memory</span> <span>=</span> <span>ReplayMemory</span><span>(</span><span>capacity</span><span>=</span><span>1000</span><span>)</span> <span># Input spaces contain spaces for all arguments.</span> <span>input_spaces</span> <span>=</span> <span>dict</span><span>(</span><span>records</span><span>=</span><span>record_space</span><span>,</span> <span>num_records</span><span>=</span><span>int</span><span>)</span> <span># Builds the memory with variables, placeholders for these spaces.</span> <span>graph</span> <span>=</span> <span>ComponentTest</span><span>(</span><span>component</span><span>=</span><span>memory</span><span>,</span> <span>input_spaces</span><span>=</span><span>input_spaces</span><span>)</span> <span># Generate a sample batch from the nested space.</span> <span>observation</span> <span>=</span> <span>record_space</span><span>.</span><span>sample</span><span>(</span><span>size</span><span>=</span><span>32</span><span>)</span> <span># Calls session, fetches ops, prepares inputs, executes API method.</span> <span>graph</span><span>.</span><span>test</span><span>(</span><span>memory</span><span>.</span><span>insert_records</span><span>,</span> <span>observation</span><span>))</span> <span># Get some samples providing the "num_records" int-arg.</span> <span>samples</span> <span>=</span> <span>graph</span><span>.</span><span>test</span><span>((</span><span>memory</span><span>.</span><span>insert_records</span><span>,</span> <span>10</span><span>))</span> </code></pre></div></div> <p>Separating spaces of tensors from logical composition enables us to reuse components without ever manually dealing with incompatible shapes again. Note how the above code does not contain any framework-specific notions but only defines an input dataflow from a set of spaces. In RLgraph, heuristics (which often have great impact on performance in RL) are not afterthoughts but first class citizens which are tested both individually and in integration with other components. For example, the <code>Policy</code> component contains neural network, action adapter (and in turn layer and distribution) sub-components, all of which are tested separately.</p> <p>The core difference between using RLgraph and standard implementation workflows is that every component is fully specified explicitly: Its devices and scopes for computations and internal states (e.g. variable sharing) are explicitly assigned which spares developers the headaches of nested context managers. As a result, RLgraph creates beautiful TensorBoard visualisations, e.g. our IMPALA implementation:</p> <p><img alt="RLgraph IMPALA tensorboard" src="https://rlgraph.github.io/images/impala_tboard_graph_learner.png"></p> <h2>Resources and call for contributions</h2> <p>RLgraph is in alpha stage and being used in a number of research pilots. We welcome contributions and feedback. As RLgraph ambitiously covers multiple frameworks and backends, there is a lot to do on all ends (we would especially welcome more PyTorch expertise).</p> <p>Over the next few months, we will keep building out utilities (TF 2.0 considerations, separation of backend code (specifically see pinned issues), additional Ray executors, …), and implement more algorithms. Feel free to create issues discussing improvements.</p> <p>Code: Get started building applications with RLgraph: <a href="https://github.com/rlgraph/rlgraph">https://github.com/rlgraph/rlgraph</a></p> <p>Documentation: Docs are available on <a href="https://rlgraph.readthedocs.io/en/latest/?badge=latest">readthedocs</a></p> <p>Paper: Read our <a href="https://arxiv.org/abs/1810.09028">paper</a> to learn more about RLgraph’s design.</p> </section></div><br>
<br>
<br>
<br>
<br>
<br>
via RLgraph <br>
<br>
March 18, 2019 at 03:01PM
